internal/venice-common/src/main/java/com/linkedin/venice/utils/AvroRecordUtils.java

```diff
@@ -0,0 +1,235 @@
+            Collection collection = (Collection) defaultValue;
+            if (collection.isEmpty()) {
+              recordType.put(field.pos(), new ArrayList<>());
+            } else {
+              throw new VeniceException(
+                  "Non 'empty array' default value is not supported for array type: " + field.name());
+            }
+            break;
+          case MAP:
+            Map map = (Map) defaultValue;
+            if (map.isEmpty()) {
+              recordType.put(field.pos(), new HashMap<>());
+            } else {
+              throw new VeniceException("Non 'empty map' default value is not supported for map type: " + field.name());
+            }
+            break;
+          case ENUM:
+          case FIXED:
+          case BYTES:
+          case RECORD:
+            throw new VeniceException(
+                "Default value for field: " + field.name() + " with type: " + fieldType + " is not supported");
+        }
+      }
+    }
+
+    return recordType;
+  }
+
+  /**
+   * Checks if it is possible for some value to be modified to adapt to the provided schema type.
+   * @param expectedSchemaType The type {@link Schema.Type} that the value needs to be adapted to.
+   * @return {@code true} if a value can be modified to adapt to the provided schema type; {@code false} otherwise.
+   */
+  private static boolean canSchemaTypeBeAdapted(Schema.Type expectedSchemaType) {
+    switch (expectedSchemaType) {
+      case ENUM:
+      case FIXED:
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
+        return false;
+      case RECORD:
+      case ARRAY:
+      case MAP:
+      case UNION:
+        return true;
+      default:
+        // Defensive coding
+        throw new VeniceException("Unhandled Avro type. This is unexpected.");
+    }
+  }
+
+  /**
+   * Tries to adapt the {@param datum} to the {@param expectedSchema}.
+   * The following steps are followed:
+   * 1. If the schema type doesn't allow for adaptation, return the same value that was passed in input.
+   * 2. If the schema type allows for adaptation, then
+   *     2a. If the value doesn't specify a value for any field, the default value is used
+   *     2b. If a field is mandatory, but no default values are specified, then an Exception is thrown
+   *
+   * @param expectedSchema The type {@link Schema} that the value needs to be adapted to.
+   * @param datum The value that needs to be adapted to the specified schema
+   * @return
+   * @throws VeniceException
+   */
+  public static Object adaptToSchema(Schema expectedSchema, Object datum) throws VeniceException {
+    Schema.Type expectedSchemaType = expectedSchema.getType();
+    switch (expectedSchemaType) {
+      case ENUM:
+      case FIXED:
+      case STRING:
+      case BYTES:
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+      case BOOLEAN:
+      case NULL:
```

Comment:
Duplicate logic from `canSchemaTypeBeAdapted`, so can we just update the caller of `canSchemaTypeBeAdapted` to always use `adaptToSchema`?

Reply:
If the reason to maintain a separate function is to improve the perf, can we reuse the same logic?

Reply:
I've tried to deduplicate this logic by making a `SchemaAdapter` interface. Let me know if this doesn't look good. I can try to find another way to dedup this